### PR TITLE
kernel.projectDir must be set when initializing the kernel

### LINKED ISSUE #28: [4.4-beta1] kernel.project_dir parameter points to the manager bundle directory
`$container->getParameter('kernel.project_dir')` returns `/var/www/project/vendor/contao/manager-bundle` because [`Kernel::getProjectDir()` searches for the nearest *composer.json* file](https://github.com/symfony/symfony/blob/e7da63a51f6250bb69f7652bd067fc0d1623e79a/src/Symfony/Component/HttpKernel/Kernel.php#L317). As this parameter is used for `TL_ROOT`, this makes Contao 4.4-beta1 unusable in the managed edition.

I think we have to add a `getProjectDir()` method to the `ContaoKernel`.

### PR DESCRIPTION
**Pull request by aschempp**
fixes #28

### CODE REVIEW

**Comment by ausi** [comment]
Couldn’t we just use `dirname(dirname(dirname(dirname(dirname(__DIR__)))))` in `getProjectDir()` and not inject the path? Symfony itself also doesn’t inject the project dir but searches for the nearest composer.json file.

**Comment by aschempp** [comment]
What if classes are combined for caching etc?

**Comment by ausi** [comment]
I think since PHP 7 there is no need for caching PHP files by combining them. But even then, the caching script should handle the `__DIR__` constants correctly.

File: src/HttpKernel/ContaoKernel.php [hunk 1]
```diff
@@ -64,6 +91,8 @@ public function getRootDir()
      * Sets the application root dir.
      *
      * @param string $dir
+     *
+     * @deprecated Deprecated since version 4.4, to be removed in 5.0.
      */
     public function setRootDir($dir)
     {
```
Comment (ausi) [inline src/HttpKernel/ContaoKernel.php]: Should we add a `@trigger_error(, E_USER_DEPRECATED)` here?

Comment (aschempp) [reply src/HttpKernel/ContaoKernel.php]: Valid point, added in b1702b93a48bd3f30fb9f6ff728fa2ee70b80da2

**Review by leofeyer — APPROVED** [review]

**Review by ausi — CHANGES_REQUESTED** [review]
If we want to go with the constructor argument, I think the following places have to be updated too:

- [src/Resources/system/initialize.php:22](https://github.com/contao/manager-bundle/blob/4d2babb216341cf46c8146fc50730028af6e8275/src/Resources/system/initialize.php#L22)
- [bin/contao-console:43](https://github.com/contao/manager-bundle/blob/4d2babb216341cf46c8146fc50730028af6e8275/bin/contao-console#L43)
- [tests/HttpKernel/ContaoKernelTest.php:53](https://github.com/contao/manager-bundle/blob/4d2babb216341cf46c8146fc50730028af6e8275/tests/HttpKernel/ContaoKernelTest.php#L53)

But I would still prefer the `dirname()` solution.

Also, isn’t changing the constructor arguments a BC break? E.g. what is if I use a custom *app_dev.php* in my project that uses the `ContaoKernel`, wouldn’t that break with an update?

**Comment by aschempp** [comment]
I have fixed the remaining uses of `ContaoKernel::setProjectDir`. Regarding BC break, afaik the constructor is not a BC promise.

**Comment by Toflar** [comment]
> Regarding BC break, afaik the constructor is not a BC promise.

Of course it is. You can only add arguments that are optional.

**Comment by leofeyer** [comment]
I have updated the PR so it is backwards compatible. Please review again.

**Comment by leofeyer** [comment]
We should also wait for https://github.com/symfony/symfony/pull/22727 before merging this PR.

File: src/HttpKernel/ContaoKernel.php [hunk 2]
```diff
@@ -51,10 +70,22 @@ public function registerBundles()
     /**
      * {@inheritdoc}
      */
+    public function getProjectDir()
+    {
+        if (null === $this->projectDir) {
+            $this->projectDir = dirname(dirname(dirname(dirname(dirname(__DIR__)))));
```
Comment (ausi) [inline src/HttpKernel/ContaoKernel.php]: With this line, the ContaoKernel now finds the path to the project dir itself. Why do we need the constructor argument then?

Comment (leofeyer) [reply src/HttpKernel/ContaoKernel.php]: For the unit tests I guess.

Comment (ausi) [reply src/HttpKernel/ContaoKernel.php]: In the unit tests we could compare the project dir path against the path of the ContaoKernel.php file.

File: src/HttpKernel/ContaoKernel.php [hunk 3]
```diff
@@ -37,6 +42,20 @@ class ContaoKernel extends Kernel
     private $bundleLoader;
 
     /**
+     * Constructor.
+     *
+     * @param string|null $projectDir
```
Comment (ausi) [inline src/HttpKernel/ContaoKernel.php]: I think `null` should not be allowed here.

**Review by ausi — APPROVED** [review]
One parameter doccomment is wrong IMO, the rest looks good to me 👍

**Comment by leofeyer** [comment]
Unfortunately, adding `$projectDir` as first argument leads to an exception upon install:

```
composer run-script post-update-cmd
> Contao\ManagerBundle\Composer\ScriptHandler::initializeApplication

 Added/updated the web/app.php file.

 // Clearing the cache for the prod environment with debug                      
 // false                                                                       

21:37:54 ERROR     [console] Error thrown while running command "--ansi cache:clear --env=prod". Message: "Too few arguments to function Contao\ManagerBundle\HttpKernel\ContaoKernel::__construct(), 2 passed in /Users/leofeyer/Sites/contao/managed-edition/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Command/CacheClearCommand.php on line 262 and exactly 3 expected" ["error" => ArgumentCountError {trace: { …} …},"command" => "--ansi cache:clear --env=prod","message" => "Too few arguments to function Contao\ManagerBundle\HttpKernel\ContaoKernel::__construct(), 2 passed in /Users/leofeyer/Sites/contao/managed-edition/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Command/CacheClearCommand.php on line 262 and exactly 3 expected"] []
21:37:54 ERROR     [console] Command "--ansi cache:clear --env=prod" exited with code "1" ["command" => "--ansi cache:clear --env=prod","code" => 1] []

  [Symfony\Component\Debug\Exception\FatalThrowableError]                      
  Type error: Too few arguments to function Contao\ManagerBundle\HttpKernel\C  
  ontaoKernel::__construct(), 2 passed in /Users/leofeyer/Sites/contao/manage  
  d-edition/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Command  
  /CacheClearCommand.php on line 262 and exactly 3 expected                    
```

This is because the Symfony [CacheClearCommand](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/Command/CacheClearCommand.php#L200) creates a temporary kernel, which of course does not know about our custom parameter.

So I guess there is no other way then making the third argument optional?

**Pull request merged**